worldedit-core/src/main/java/com/sk89q/worldedit/util/schematic/backends/PollingSchematicsBackend.java

```diff
@@ -0,0 +1,108 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.sk89q.worldedit.util.schematic.backends;
+
+import com.sk89q.worldedit.internal.util.LogManagerCompat;
+import com.sk89q.worldedit.util.schematic.SchematicPath;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * SchematicsBackend implementation that scans the folder tree, then caches the result for a certain amount of time.
+ * This essentially is an eventually consistent cache that is used as fallback.
+ */
+public class PollingSchematicsBackend implements SchematicsBackend {
+
+    private static final Logger LOGGER = LogManagerCompat.getLogger();
+
+    private static final Duration MAX_RESULT_AGE = Duration.ofSeconds(10);
+
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    private final Path schematicsDir;
+    private Instant lastUpdateTs = Instant.EPOCH;
+    private List<SchematicPath> schematics = new ArrayList<>();
+
+    private PollingSchematicsBackend(Path schematicsDir) {
+        this.schematicsDir = schematicsDir;
+    }
+
+    /**
+     * Create a new instance of the polling SchematicsManager backend.
+     * @param schematicsFolder Root folder for schematics.
+     * @return A new PollingSchematicsBackend instance.
+     */
+    public static PollingSchematicsBackend create(Path schematicsFolder) {
+        return new PollingSchematicsBackend(schematicsFolder);
+    }
+
+    private List<SchematicPath> scanFolder(Path root) {
+        List<SchematicPath> pathList = new ArrayList<>();
+        try (DirectoryStream<Path> stream = Files.newDirectoryStream(root)) {
+            for (Path path : stream) {
+                if (Files.isDirectory(path)) {
+                    pathList.addAll(scanFolder(path));
+                } else {
+                    pathList.add(new SchematicPath(path));
+                }
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return pathList;
+    }
+
+    private void runRescan() {
+        LOGGER.debug("Rescanning Schematics");
+        this.schematics = scanFolder(schematicsDir);
+        lastUpdateTs = Instant.now();
+    }
+
+    @Override
+    public void init() {
+    }
+
+    @Override
+    public void uninit() {
+    }
+
+    @Override
+    public synchronized List<SchematicPath> getList() {
+        // udpate internal cache if requried (determined by age)
+        Duration age = Duration.between(lastUpdateTs, Instant.now());
+        if (age.compareTo(MAX_RESULT_AGE) >= 0) {
+            runRescan();
+        }
```

Comment:
I wonder if we shouldn't be doing this more proactively, using a scheduled executor service. Maybe if there's significant lag on this for someone, we can revisit.